code/modules/mafia/roles.dm

```diff
@@ -0,0 +1,486 @@
+	var/datum/mafia_role/R = current_target
+	if(R)
+		to_chat(body,"<span class='warning'>You invoke spirit of [R.body.real_name] and learn their role was <b>[R.name]<b>.</span>")
+		add_note("N[game.turn] - [R.body.real_name] - [R.name]")
+		current_target = null
+
+/datum/mafia_role/lawyer
+	name = "Lawyer"
+	desc = "You can choose a person during the day to provide extensive legal advice to during the night, preventing night actions."
+
+	revealed_outfit = /datum/outfit/mafia/lawyer
+	targeted_actions = list("Advise")
+
+	var/datum/mafia_role/current_target
+
+/datum/mafia_role/lawyer/New(datum/mafia_controller/game)
+	. = ..()
+	RegisterSignal(game,COMSIG_MAFIA_SUNDOWN,.proc/roleblock_text)
+	RegisterSignal(game,COMSIG_MAFIA_NIGHT_START,.proc/try_to_roleblock)
+	RegisterSignal(game,COMSIG_MAFIA_NIGHT_END,.proc/release)
+
+/datum/mafia_role/lawyer/proc/roleblock_text(datum/mafia_controller/game)
+	if(SEND_SIGNAL(src,COMSIG_MAFIA_CAN_PERFORM_ACTION,game,"roleblock",current_target) & MAFIA_PREVENT_ACTION || game_status != MAFIA_ALIVE) //Got lynched or roleblocked by another lawyer.
+		current_target = null
+	if(current_target)
+		to_chat(current_target.body,"<span class='big bold red'>YOU HAVE BEEN BLOCKED! YOU CANNOT PERFORM ANY ACTIONS TONIGHT.</span>")
+		add_note("N[game.turn] - [current_target.body.real_name] - Blocked")
+
+/datum/mafia_role/lawyer/validate_action_target(datum/mafia_controller/game, action, datum/mafia_role/target)
+	. = ..()
+	if(!.)
+		return FALSE
+	if(game.phase == MAFIA_PHASE_NIGHT)
+		return FALSE
+	if(target.game_status != MAFIA_ALIVE)
+		return FALSE
+
+/datum/mafia_role/lawyer/handle_action(datum/mafia_controller/game, action, datum/mafia_role/target)
+	. = ..()
+	if(target == current_target)
+		current_target = null
+		to_chat(body,"<span class='warning'>You have decided against blocking anyone tonight.</span>")
+	else
+		current_target = target
+		to_chat(body,"<span class='warning'>You will block [target.body.real_name] tonight.</span>")
+
+/datum/mafia_role/lawyer/proc/try_to_roleblock(datum/mafia_controller/game)
+	if(current_target)
+		RegisterSignal(current_target,COMSIG_MAFIA_CAN_PERFORM_ACTION, .proc/prevent_action)
+
+/datum/mafia_role/lawyer/proc/release(datum/mafia_controller/game)
+	. = ..()
+	if(current_target)
+		UnregisterSignal(current_target)
+		current_target = null
+
+/datum/mafia_role/lawyer/proc/prevent_action(datum/source)
+	if(game_status == MAFIA_ALIVE) //in case we got killed while imprisoning sk - bad luck edge
+		return MAFIA_PREVENT_ACTION
+
+/datum/mafia_role/psychologist
+	name = "Psychologist"
+	desc = "You can visit a single member of the town to reveal their true role in the morning!"
+	revealed_outfit = /datum/outfit/mafia/psychologist
+
+	targeted_actions = list("Reveal")
+	var/datum/mafia_role/current_target
+	var/can_use = TRUE
+
+/datum/mafia_role/psychologist/New(datum/mafia_controller/game)
+	. = ..()
+	RegisterSignal(game,COMSIG_MAFIA_NIGHT_START,.proc/try_to_visit)
```

Comment:
Why at night start. If they get killed it will still go through.

Reply:
great point